accepted/00000-replace-salt-toaster.md

```diff
@@ -0,0 +1,192 @@
+- Start Date: 2022-08-03
+
+# Summary
+[summary]: #summary
+
+Replace Salt Toaster with an approach that's closer to upstream's testing efforts.
+
+# Motivation
+[motivation]: #motivation
+
+<!-- - Why are we doing this? -->
+<!-- - What use cases does it support? -->
+<!-- - What is the expected outcome? -->
+
+<!-- Describe the problem you are trying to solve, and its constraints, without coupling them too closely to the solution you have in mind. If this RFC is not accepted, the motivation can be used to develop alternative solutions. -->
+
+Salt Toaster was developed a few years ago and filled a gap in the Salt world.
+Over the last years, maintaining the downstream `suse.tests` was too much effort
+-- so much that we didn't do it. Maintaining the docker images that Salt Toaster
+uses to run the upstream test suite against our patched Salt grew into a bigger
+and bigger task. We didn't keep up with upstream's changes to the test
+dependencies well and use a number of workarounds to still run a test suite.
+
+At the same time, Salt upstream improved their test suite a lot. They added new
+test types and now have a similar test framework to Salt Toaster. 
+
+The purpose of this change is to lessen our maintenance burden and to be good
+open-source citizens that collaborate with our upstream project when it comes to
+testing Salt. Additionally, we have knowledge about running a test suite within
+our team that we want to take advantage of.
+
+# Detailed design
+[design]: #detailed-design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the product to understand, and for somebody familiar with the internals to implement. -->
+
+<!-- This section should cover architecture aspects and the rationale behind disruptive technical decisions (when applicable), as well as corner-cases and warnings. Whenever the new feature creates new user interactions, this section should include examples of how the feature will be used. -->
+
+## Overview
+
+The Salt upstream test suite is executed in virtual machines (VMs) that run the
+different operating systems we support. These VMs are set up with sumaform and
+salt-jenkins. Terracumber triggers terraform to create the VMs from within a
+Jenkins job and launches the test suite, connecting to each VM over SSH.
+
+## Provisioning Test VMs with Sumaform + Salt-Jenkins
+
+A new module "salt\_testenv" is added to sumaform. This module defines a new
+role, "salt\_testenv", and the default values needed to run all test suites, e.g.
+4GiB of RAM. Sumaform uses this role to apply the `golden-images-provision.sls`
+state file from the Salt-Jenkins repository.
+
+Up to now, all state files used by sumaform are maintained in the sumaform Git
+repository. For provisioning the salt\_testenv VMs, we don't want to maintain
+our own states that define which dependencies need to be installed on what
+operating system. Salt upstream already maintains such states for their own test
```

Comment:
Hmmm, I'm not sure here, as one of the thing the current Salt toaster was taking care of is to test the installation of the Salt package (RPM or DEB) that we built.

Also run the tests using the Salt dependencies that are installed also from the OS + custom Salt repos we built.

This way, we ensure we are testing the Salt codebase with the actual dependencies and versions that customer are going to get when installing this Salt package in real life.

Do you have any idea how to solve this?

If we rely on Salt upstream and install the Salt dependencies via "pip", with a predefined version specified by upstream, there might be differences with the versions the customers will get when install Salt coming from our repos, or that some of the dependencies are not even available in these repos.

So, ideally during provisioning, we would like to test the package installation of our generated Salt package, and then while testing, using the dependencies that are pulled as requires when installing the Salt package.

This way we could cover the cases of: 
- Missing installation dependencies for Salt in our repositories. 
- The version of an installed RPM/DEB dependency of Salt is different that what "pip" installs. (due available packages in repo or due spec/control speficiation)

Reply:
That's quite a big comment, let me try to clarify this a bit.

> Hmmm, I'm not sure here, as one of the things the current Salt toaster was taking care of is to test the installation of the Salt package (RPM or DEB) that we built.
> 

We still need to put the Salt source code on the test VM, I didn't cover that yet. Salt should come from our packages to ensure we test the program we want to test. That this will also act as an "installation test" is a nice side effect, but IMO not really important within the scope of running Salt's testsuite. 


> Also runs the tests using the Salt dependencies that are installed also from the OS + custom Salt repos we built.
> 

That stays the same.

> This way, we ensure we are testing the Salt codebase with the actual dependencies and versions that customers are going to get when installing this Salt package in real life from our repos.
> 
> Do you have any idea how to solve this?
> 

I haven't looked deeply into this, right now I expect we can install Salt from the correct repository with `terracumber` just ahead of executing the test suite. 

> If we rely on Salt upstream and install the Salt dependencies via "pip", 

The Salt state tree that provisions the VMs does not install Salt dependencies with `pip`, with the exception if two state files. We need to decide what to do about these two.

>with a predefined version specified by upstream, there might be differences with the versions the customers will get when installing Salt coming from our repos, or that some of the dependencies are not even available in these repos.
> 
> So, ideally during provisioning, we would like to test the package installation of our generated Salt package, and then while testing, using the dependencies that are pulled as requires when installing that Salt package.
>

I agree that this would be ideal. At this point, we can't enforce that the dependencies used during the test execution come from packages because `nox` uses `pip` to fill the virtual environment used for testing. I don't know how much work it would be to override that behavior. 

> This way we could cover the cases of:
> 
>     * Missing installation dependencies for Salt in our repositories.
> 
>     * The version of an installed RPM/DEB dependency of Salt is different that what "pip" installs. (due available packages in repo or due spec/control speficiation)
> 
> 
> What do you think?

I think we need to consider the trade-off between how closely the test environment resembles the production environment and how much work it is to maintain the test infrastructure. If it turns out it's not possible to use the very same dependencies in the test environment but we see that as a requirement, we would need to rethink our approach to testing without `nox` or using `pip`.

Reply:
We discussed this topic a bit on a call. The outcome is that we evaluate if `nox` actually makes running the tests easier (against our packaged dependencies). The alternative to compare this against is to run a plain `pytest`. In that case we need to install all dependencies used in the tests manually. A `python3X-salt-tests` subpackage could require all the dependencies (that are for tests only) and ship the tests themselves.